MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,322 @@
+    /// It will return none if the hash does not exist
+    pub fn get_mut_object(&mut self, hash: ObjectHash) -> Option<&mut T> {
+        self.data.get_mut(&hash)
+    }
+
+    /// This function returns the hash proof tree of a given hash.
+    /// If the given hash is not in the tree, the vec will be empty.
+    /// The Vec will be created in form of the child-child-parent
+    pub fn get_hash_proof(&self, hash: ObjectHash) -> Vec<ObjectHash> {
+        let mut result = Vec::new();
+        let mut i = self.mmr.len();
+        for counter in 0..self.mmr.len() {
+            if self.mmr[counter].hash == hash {
+                result.push(self.mmr[i].hash.clone());
+                i = counter;
+                break;
+            }
+        }
+        i = peer_index(i as u64) as usize;
+        while i < self.mmr.len() {
+            result.push(self.mmr[i].hash.clone());
+            i += 1;
+            result.push(self.mmr[i].hash.clone());
+
+            i = peer_index(i as u64) as usize;
+        }
+        result
+    }
+
+    /// This function returns the peak height of the mmr
+    pub fn get_peak(&self) -> u32 {
+        let mut height_counter = 0;
+        let mmr_len = self.get_last_added_index() as i128;
+        while (mmr_len - i128::pow(2, height_counter + 1) - 2) >= 0 {
+            height_counter += 1;
+        }
+        height_counter
+    }
+
+    /// This function adds a vec of leaf nodes to the mmr.
+    pub fn add_vec(&mut self, objects: Vec<T>) {
+        for object in objects {
+            self.add_single(object);
+        }
+    }
+
+    /// This function adds a new leaf node to the mmr.
+    pub fn add_single(&mut self, object: T) {
+        let node_hash = object.get_hash();
+        let node = MerkleNode::new(node_hash);
+        self.mmr.push(node);
+        if is_node_right(self.get_last_added_index()) {
+            self.data.insert(node_hash, object);
+            self.add_single_no_leaf(self.get_last_added_index())
+        }
+    }
+
+    // This function adds non leaf nodes, eg nodes that are not directly a hash of data
+    // This is iterative and will continue to up and till it hits the top, will be a future left child
+    fn add_single_no_leaf(&mut self, index: u64) {
+        let new_hash =
+            self.data.get(&self.mmr[index as usize].hash).unwrap().concat(self.mmr[peer_index(index) as usize].hash);
+
+        let new_node = MerkleNode::new(new_hash);
+        self.mmr.push(new_node);
+        if is_node_right(self.get_last_added_index()) {
+            self.add_single_no_leaf(self.get_last_added_index())
+        }
+    }
+
+    // This function is just a private function to return the index of the last added node
+    fn get_last_added_index(&self) -> u64 {
+        (self.mmr.len() - 1) as u64
+    }
+}
+/// This function takes in the index and calculates the index of the peer.
+pub fn peer_index(index: u64) -> u64 {
```

Comment:
Prefer keeping this in the struct as a class function.

Reply:
I looked at this, but because the mmr is of uses <T, D>, you have to supply it the type of T and D if you call those functions. Thos functions dont really care about T and D.

Reply:
a Call to that function in side the  mmr looks like this:
            i = MerkleMountainRange::<T, D>::peer_index(i);

Reply:
Currently its called like this:
a Call to that function in side the  mmr looks like this:
`i = peer_index(i);`

a Call outside will look like this:
`        i = MerkleMountainRange::peer_index(i);`